File: Best Time to Buy and Sell Stock II - Leetcode 122/Best Time to Buy and Sell Stock II - Leetcode 122.java
class Solution {
    public int maxProfit(int[] prices) {
        int i = 0;
        int lo = prices[0];
        int hi = prices[0];
        int profit = 0;
        int n = prices.length;

        while (i < n - 1) {
            // Look where to buy
            while (i < n - 1 && prices[i] >= prices[i + 1]) {
                i++;
            }
            lo = prices[i];

            // Look where to sell
            while (i < n - 1 && prices[i] <= prices[i + 1]) {
                i++;
            }
            hi = prices[i];

            profit += hi - lo;
        }

        return profit;
    }
}

File: Best Time to Buy and Sell Stock II - Leetcode 122/Best Time to Buy and Sell Stock II - Leetcode 122.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i = 0
        lo = prices[0]
        hi = prices[0]
        profit = 0
        n = len(prices)

        while i < n-1:
            # look where to buy
            while i < n-1 and prices[i] >= prices[i+1]:
                i += 1
            lo = prices[i]

            # look where to sell
            while i < n-1 and prices[i] <= prices[i+1]:
                i += 1
            hi = prices[i]
            
            profit += hi - lo
        
        return profit # Time: O(n), Space: O(1)